### ISSUE TITLE
Mesh Bed Leveling w/ ABL not working

### ISSUE THREAD

**Issue by backpocketstudios** [KevinOConnor/klipper#1196]
I'm having some issues getting mesh bed leveling to work. I've attached a M112 Klippy log for reference. I have tried saving a mesh with "Save_config" and this run has no mesh saved and is probing just before printing. I've set the z-min to a negative number that should be just outside of my probe points as recommended, I've also manually calibrated the z-offset several times.

I now have my Gcode starting script running the gcode_macro you see in the config to mesh level just before printing (thought perhaps homing was overriding the mesh).

I am a web developer, and I'm not afraid to get my hands dirty with python, but I don't see any logical reason why this simple mesh config doesn't work (I've wiped out my more detailed config to see if that was perhaps the issue).

I welcome any and all suggestions, or requests for further info/logs. A simple benchy test print in the center of my CR10S bed worked, but anything along the edges (that have significant droop) has issues. This worked flawlessly when running the EZABL version of Marlin.

Thank you.
[klippy.zip](https://github.com/KevinOConnor/klipper/files/2830368/klippy.zip)

**Comment by Arksine** [comment]
To be clear, you are having issues printing in the area outside of the probing area?  In your example, over 120mm of the bed along the x-axis is not probed (62 mm on the left side, 63 mm on the right side).  I don't know what TH3D Marlin, does, but Klipper's mesh leveling clamps adjustment to the edges of the mesh.  Perhaps the TH3D firmware does some sort of extrapolation outside of the mesh, Klipper does not do this.

Ideally to get a more accurate mesh with more bed usage you would want to probe a larger section of the bed with more probe points.  Unfortunately the large x_offset of your probe is a limiting factor.

**Comment by backpocketstudios** [comment]
Hmm that would explain some of the issues, if I were to print in an area
outside the probed area and then move inside the mesh, would the mesh then
kick in or would it be ignored for the entire print?

I ask because the leveling test I am printing starts near the edge (far
enough to possibly be outside the probed area) but then moves to near
center (well within the probed area). The calibration boxes are clearly not
adjusting for z-height at any point.

Thank you for the suggestion, I will try to expand my probed area when I
get home from work tonight.
Chad Heiser

**Comment by Arksine** [comment]
Yes, when the tool moves back inside the mesh Z will be adjusted according the the value inside the mesh.

To be clear, according to your current configuration, the mesh overlays the following coordinates:

(188,312)----------------(363,312)
|                                                  |
|                                                  |
|                                                  |
|                                                  |
|                                                  |
(188,32)------------------(363,32)

I sincerely doubt that is your intention if the bed is 300x300.  I suspect both your x and y offsets in [probe] should be negative numbers, otherwise you would be probing off of the bed.  If the probe is located in the negative direction on an axis relative to the nozzle, then it needs a negative offset.

If your offsets aren't configured correctly then the mesh won't be properly mapped to the tool's coordinates.  This would result in incorrect Z adjustment across the entire bed.

**Comment by backpocketstudios** [comment]
Ok thank you very much. I will verify all this when I get back to the
machine tonight.

I have the mesh section set more on the "safe" side or so I thought.

If I set the probe offsets incorrectly, would the mesh procedure attempt to
force the nozzle/probe off the bed? I had originally set the Y offset (if I
recall) negative and I had an issue with the nozzle wanting to move off the
bed and therefore triggering an error until I flipped it. Again I will test
and verify again once I am home. My current setup has the probe held quite
offset from the nozzle with a fang cooler and mount. I have a new Titan
Aero and Volcano waiting to be installed but I wanted to hammer out any
issues with Klipper first before introducing more variables.

Again thank you, I appreciate the insight.
Chad Heiser

**Comment by Arksine** [comment]
The offsets in [probe] don't affect movement during the probing procedure.   The points themselves are determined by [bed_mesh], using the min_point, max_point, and probe_count parameters.  If the probe moved off of the bed it was likely because y coordinate of min point was too low.

**Comment by linux-paul** [comment]
Maybe I'm wrong I didn't print outside the mesh till jet, but is it possible that I've seen differences in mesh values regarding to to the z speed value? Also homing. and tilting. in my case a higher z speed results in a lower nozzle pos. In my opinion the mesh offset at home pos should be close to zero and not ~0.1mm. After I used a probe speed of 2.5 I got nice mesh values. (~0.01mm @ home pos)

**Comment by Arksine** [comment]
@linux-paul That all depends on the accuracy of your probe.  Slowing down the speed could help make probing results more repeatable.

**Comment by linux-paul** [comment]
yes, but the differences in just one direction feel a little bit strange, but I'm not a pro. During my next tests I'd like to reduce the homing speed. Thank You :)

**Comment by backpocketstudios** [comment]
Hello,

Eric, thank you for all your help. You were absolutely correct, my offsets
for my probe should have been negative, as well as my mesh_min/max needed
to be pushed as close to the edge as possible. I then re-calibrated my
z-offset for the probe and I got promising results. I'm still missing 70mm
on the x on the left side and about the same on the right, but it's better
than 120mm, this is simply due to my mount and how far out it pushes my
probe.

To be honest, I haven't had a chance to run a full test, and the CR10 bed
test was too wide to get a good run with that, but I've been wrapped up
with work and haven't had the proper time to work on it.

I've slowed my probing to 2.5 from 5 but it doesn't seem to offer that much
improvement. My min deflection from "zero" is still nearly 1mm which is
concerning.

It appears I just have more fiddling to do to get the mesh working
entirely, and when I change my hotend/extruder I need to work on getting
the probe as close to the nozzle as I can.

Again I appreciate the help.
Chad Heiser
